Escape behavior should match that of Branch List

iAmWillShepherd opened the pull request:
Fixes issue #3597

## Before
_see original issue for gif_
## After
![fix](https://user-images.githubusercontent.com/1715082/35988665-b6d607d8-0cc4-11e8-81d5-c8609b0c4ffb.gif)

tierninho commented:
✔️  looks good

shiftkey commented on app/src/ui/branches/branches-container.tsx:
```diff
@@ -65,17 +67,23 @@ export class BranchesContainer extends React.Component<
   }
 
-  private onFilterKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
+  private closeFoldoutOnEsc = (when: () => boolean) => (
```
:lipstick: could this be named to something more descriptive? `shouldCloseFoldout`?

shiftkey replied:
Sorry, I actually meant the `when` parameter here, not the function name 🤦‍♂️

shiftkey commented on app/src/ui/branches/branches-container.tsx:
```diff
@@ -65,17 +67,23 @@ export class BranchesContainer extends React.Component<
   }
 
-  private onFilterKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
+  private shouldCloseFoldout = (when: () => boolean) => (
```
Sorry, I actually meant the `when` parameter here, not the function name 🤦‍♂️ 

`onFilterKeyDown` here was fine.

iAmWillShepherd replied:
Where's the Homer Simpson Doh! when you need it?! 😆

iAmWillShepherd commented:
🔢

niik commented on app/src/ui/branches/branches-container.tsx:
```diff
@@ -134,9 +142,11 @@ export class BranchesContainer extends React.Component<
-            onFilterTextChanged={this.onFilterTextChanged}
+            filterText={this.state.branchFilterText}
+            onFilterKeyDown={this.closeFoldoutOnEsc(
+              () => this.state.branchFilterText.length === 0
```
Having anonymous functions inline like this means that the component will always re-render since every time this function is called a new instance of this function will be instantiated. We have a [tslint rule](https://github.com/desktop/desktop/blob/a991d2a0469120b843e06950add087fd6a2be2d2/tslint-rules/reactNoUnboundDispatcherPropsRule.ts) to prevent this but it only appears to work from within `render()`

This should be `onFilterKeyDown={this.closeFoldoutOnEsc}` and `closeFoldoutOnEsc` should be a bound method (i.e. `=>`) so that it's able to access this.state

j-f1 commented on app/src/ui/branches/branches-container.tsx:
```diff
@@ -181,9 +184,7 @@ export class BranchesContainer extends React.Component<
-          onFilterKeyDown={this.closeFoldoutOnEsc(
-            () => this.state.pullRequestFilterText.length === 0
-          )}
+          onFilterKeyDown={this.onPullRequestFilterKeyDown()}
```
`{this.onPullRequestFilterKeyDown}`?

j-f1 replied:
oh, I see how you’re using this. That still ends up creating a new function in every render.

iAmWillShepherd replied:
Suggestions to prevent this? I know I can duplicate the code, but was hoping to avoid that.

j-f1 replied:
You could do `onPullRequestFilterKeyDown = this.closeFoldoutOnEsc(...)` in the class body instead of defining it as a method.

iAmWillShepherd replied:
Ahh, so like a field. Seems simple enough 👍

iAmWillShepherd replied:
Linter wouldn't allow me to do that, so I just duplicated code.

iAmWillShepherd commented:
🔢

iAmWillShepherd commented:
I merged the branch `show-no-pr` into this PR to avoid confusion on my part. 🙇 @shiftkey for reviewing the PR I merged into this one.

shiftkey commented on app/src/ui/branches/branches-container.tsx:
```diff
@@ -163,39 +181,37 @@ export class BranchesContainer extends React.Component<
+        onCreatePullRequest={this.onCreatePullRequest}
+        filterText={this.state.pullRequestFilterText}
+        onFilterTextChanged={this.onPullRequestFilterTextChanged}
+        onFilterKeyDown={this.closePullRequrstFoldoutOnEsc}
```
Typo: `Requrst`

niik commented on app/src/ui/branches/branches-container.tsx:
```diff
@@ -134,9 +152,9 @@ export class BranchesContainer extends React.Component<
-            onFilterKeyDown={this.onFilterKeyDown}
-            onFilterTextChanged={this.onFilterTextChanged}
+            filterText={this.state.branchFilterText}
+            onFilterKeyDown={this.closeBranchFoldoutOnEsc}
```
I think this handler should be attached to the entire container. You should be able to hit escape while navigating the list of PRs as well as in the filter text.

iAmWillShepherd replied:
Do you mean attaching it to the `TabBar` component? If so, one of the things I struggled with was adding an event handler to plain Dom node and getting it to respond to React's synthetic events.

niik replied:
No I mean we should attach it to the branches-container div. So something like

```tsx
<div className="branches-container" onKeyDown={...}>
```

And the in the handler we'd check whether the event was handled or and if the key was 'escape'. The next change would be to have `TextBox` handle `Escape` when `type=search` and it's not empty. That way we can keep the logic specific to search inputs contained within the text component and everyone else that needs to handle escape can do their thing without having to take search boxes into account. We could take a look at this together next week if you'd like

niik commented on app/src/ui/branches/branches-container.tsx:
```diff
@@ -65,17 +66,34 @@ export class BranchesContainer extends React.Component<
+  private closePullRequrstFoldoutOnEsc = (
+    event: React.KeyboardEvent<HTMLElement>
+  ) => {
+    if (event.key === 'Escape') {
```
We should check that no handler has taken care of this before us by doing `if (!event.defaultPrevented && e.key ===`

iAmWillShepherd commented:
The current fix for this bug introduces a new issue: race conditions. This is due to our reliance on props' values in the filter list. 

@niik suggested we refactor so that we can use the values from the DOM element itself. I will be doing the following in separate PRs:
* Remove `onChange` event in favor of `onValueChanged` in`TextBox` component
* Stop leaking implementation details from `TextBox` by removing 'onInputRef` props and exposing two new public methods, `focus` and `selectAll`

niik commented:
#4077 and #4080 has landed, let's merge them in here and bring it all together!

j-f1 commented:
Before this gets merged, can you fix the typo in the PR title? (b**a**havior → b**e**havior)

iAmWillShepherd commented:
🔢

niik commented on app/src/ui/lib/text-box.tsx:
```diff
@@ -177,6 +177,14 @@ export class TextBox extends React.Component<ITextBoxProps, ITextBoxState> {
     )
   }
 
+  private onKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
```
Now that we've moved TextBox to not expose the onChange event directly I think we should move the Escape check in here instead.

```ts
if (event.key === 'Escape' && this.props.type === 'search') {
  if (this.inputElement && this.inputElement.value.length > 0) {
    event.preventDefault()
    this.props.onValueChanged('')
  }
}
```

j-f1 replied:
`onEscape`?

iAmWillShepherd replied:
I initially was going to put it in that component, then I thought about the reason why we use `Esc` to clear the contents of the textbox and decided it should be in the filter list. It seems that we only want to `Esc`to clear the contents of in filter-lists and no where else. Is that an incorrect assumption on my part?

niik replied:
@iAmWillShepherd The way I look at it is it's a behavior of a search input field. Like, if we were to use the `TextBox` with `type=search` somewhere outside of a filter list we'd still want to clear its text when you hit escape. It just so happens to be the case that we only use search textboxes inside of filter lists at the moment.

> onEscape

@j-f1 We try to match event handler names to event names so it should be `onKeyDown`

j-f1 replied:
@niik I was suggesting having an event that only gets called when the escape key is pressed.

niik commented on app/src/ui/lib/text-box.tsx:
```diff
@@ -177,6 +177,14 @@ export class TextBox extends React.Component<ITextBoxProps, ITextBoxState> {
+  private onKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
+    const eventHandler = this.props.onKeyDown
+
+    if (eventHandler != null) {
```
Can we skip the intermediate variable here? Doesn't seem to be giving us much over repeating it in the if-statement and the method call

iAmWillShepherd commented:
🔢

j-f1 commented on app/src/ui/lib/text-box.tsx:
```diff
@@ -178,6 +178,21 @@ export class TextBox extends React.Component<ITextBoxProps, ITextBoxState> {
+      this.props.type === 'search' &&
+      event.key === 'Escape'
+    ) {
+      const value: string = ''
```
I don’t think you need the `: string` here.

iAmWillShepherd replied:
I didn't either, I put it there to appease VSCode. I'll try restarting my editor to see if it still complains when I remove the type annotation.

iAmWillShepherd commented:
All good @j-f1, turns out a simple restart was all that I needed.

niik commented on app/src/ui/lib/filter-list.tsx:
```diff
@@ -340,7 +341,14 @@ export class FilterList<T extends IFilterListItem> extends React.Component<
 
-    if (event.key === 'ArrowDown') {
+    const filterText = this.props.filterText
+    if (key === 'Escape' && (filterText && filterText.length !== 0)) {
```
This section would appear to be redundant now since it's handled in the TextBox. Could we remove it please?

niik commented on app/src/ui/lib/text-box.tsx:
```diff
@@ -177,6 +177,27 @@ export class TextBox extends React.Component<ITextBoxProps, ITextBoxState> {
+      }
+    }
+
+    if (this.props.onKeyDown != null) {
```
Could we make this `(this.props.onKeyDown !== undefined)` since it's typed as undefined or function or just `if(this.props.onKeyDown)`?

niik commented on app/src/ui/branches/branches-container.tsx:
```diff
@@ -34,58 +33,36 @@ interface IBranchesProps {
   readonly selectedPullRequest: PullRequest | null
-  readonly filterText: string
+  readonly branchFilterText: string
+  readonly pullRequestFilterText: string
```
I really like that we're lifting this state into BranchesContainer

iAmWillShepherd commented:
> Remove the onDismissed prop from clone-repository.tsx

@niik , I've addressed all points except ☝️. We use that `onDismissed` prop to close the dialog. I did remove it from `CloneGithubRepository` thought.

🔢
